From 40e50c9024c4b0072b2a219fc1f15317de49ee26 Mon Sep 17 00:00:00 2001
From: "Gary D. Gregory" <ggregory@apache.org>
Date: Sat, 12 May 2012 20:33:44 +0000
Subject: [PATCH] [VFS-296][FTP] FTP socket timeout setting doesn't work if
 connect hangs. [VFS-313][FTP] Configuration does not include option for
 setting socket timeout.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/vfs/trunk@1337638 13f79535-47bb-0310-9956-ffa450edef68
---
 .../vfs2/provider/ftp/FtpClientFactory.java   |  8 ++++-
 .../ftp/FtpFileSystemConfigBuilder.java       | 30 ++++++++++++++++++-
 .../ftp/test/FtpProviderTestCase.java         |  1 +
 src/changes/changes.xml                       | 10 +++++--
 4 files changed, 45 insertions(+), 4 deletions(-)

diff --git a/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpClientFactory.java b/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpClientFactory.java
index f828d19a3e..b61e05e192 100644
--- a/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpClientFactory.java
+++ b/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpClientFactory.java
@@ -18,7 +18,6 @@
 
 import java.io.IOException;
 
-import org.apache.commons.net.ftp.FTP;
 import org.apache.commons.net.ftp.FTPClient;
 import org.apache.commons.net.ftp.FTPClientConfig;
 import org.apache.commons.net.ftp.FTPReply;
@@ -80,6 +79,13 @@ public static FTPClient createConnection(String hostname, int port, char[] usern
 
             try
             {
+                // Set connect timeout
+                Integer connectTimeout = FtpFileSystemConfigBuilder.getInstance().getConnectTimeout(fileSystemOptions);
+                if (connectTimeout != null)
+                {
+                    client.setDefaultTimeout(connectTimeout.intValue());
+                }
+
                 client.connect(hostname, port);
 
                 int reply = client.getReplyCode();
diff --git a/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpFileSystemConfigBuilder.java b/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpFileSystemConfigBuilder.java
index 4b2fc0262c..eb41507f15 100644
--- a/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpFileSystemConfigBuilder.java
+++ b/core/src/main/java/org/apache/commons/vfs2/provider/ftp/FtpFileSystemConfigBuilder.java
@@ -30,6 +30,7 @@ public final class FtpFileSystemConfigBuilder extends FileSystemConfigBuilder
 
     private static final FtpFileSystemConfigBuilder BUILDER = new FtpFileSystemConfigBuilder();
 
+    private static final String CONNECT_TIMEOUT = _PREFIX + ".CONNECT_TIMEOUT";
     private static final String DATA_TIMEOUT = _PREFIX + ".DATA_TIMEOUT";
     private static final String DEFAULT_DATE_FORMAT = _PREFIX + ".DEFAULT_DATE_FORMAT";
     private static final String ENCODING = _PREFIX + ".ENCODING";
@@ -65,6 +66,18 @@ protected Class<? extends FileSystem> getConfigClass()
         return FtpFileSystem.class;
     }
 
+    /**
+     * Gets the timeout in milliseconds to use for the socket connection.
+     * 
+     * @param opts The FileSystemOptions.
+     * @return The timeout in milliseconds to use for the socket connection.
+     * @since 2.1
+     */
+    public Integer getConnectTimeout(FileSystemOptions opts)
+    {
+        return (Integer) getParam(opts, CONNECT_TIMEOUT);
+    }
+
     /**
      * @param opts The FileSystemOptions.
      * @return The encoding.
@@ -200,6 +213,21 @@ public Boolean getUserDirIsRoot(FileSystemOptions opts)
         return getBoolean(opts, USER_DIR_IS_ROOT);
     }
 
+    /**
+     * Sets the timeout for the initial control connection.
+     * <p>
+     * If you set the connectTimeout to <code>null</code> no connectTimeout will be set.
+     * </p>
+     * 
+     * @param opts The FileSystemOptions.
+     * @param connectTimeout the timeout value in milliseconds
+     * @since 2.1
+     */
+    public void setConnectTimeout(FileSystemOptions opts, Integer connectTimeout)
+    {
+        setParam(opts, CONNECT_TIMEOUT, connectTimeout);
+    }
+
     /**
      * see {@link org.apache.commons.net.ftp.FTP#setControlEncoding} for details and examples.
      * @param opts The FileSystemOptions.
@@ -333,7 +361,7 @@ public void setShortMonthNames(FileSystemOptions opts, String[] shortMonthNames)
     }
 
     /**
-     * set the socket timeout for the ftp client.<br />
+     * Sets the socket timeout for the FTP client.<br />
      * If you set the socketTimeout to <code>null</code> no socketTimeout will be set on the
      * ftp client.
      *
diff --git a/core/src/test/java/org/apache/commons/vfs2/provider/ftp/test/FtpProviderTestCase.java b/core/src/test/java/org/apache/commons/vfs2/provider/ftp/test/FtpProviderTestCase.java
index ba98fbf0e5..d5476e8a87 100644
--- a/core/src/test/java/org/apache/commons/vfs2/provider/ftp/test/FtpProviderTestCase.java
+++ b/core/src/test/java/org/apache/commons/vfs2/provider/ftp/test/FtpProviderTestCase.java
@@ -171,6 +171,7 @@ public FileObject getBaseTestFolder(final FileSystemManager manager) throws Exce
         FtpFileSystemConfigBuilder.getInstance().setPassiveMode(opts, true);
         // FtpFileType.BINARY is the default
         FtpFileSystemConfigBuilder.getInstance().setFileType(opts, FtpFileType.BINARY);
+        FtpFileSystemConfigBuilder.getInstance().setConnectTimeout(opts, 10000);
         return manager.resolveFile(uri, opts);
     }
 
diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index cdadf89ce2..c2bf86be53 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -23,10 +23,16 @@
 
   <body>
     <release version="2.1" date="TBD" description="">
-      <action issue="VFS-414" dev="ggregory" type="update" due-to="ggregory">
+      <action issue="VFS-296" dev="ggregory" type="add" due-to="andreasp">
+	    [FTP] FTP socket timeout setting doesn't work if connect hangs.
+      </action>
+      <action issue="VFS-313" dev="ggregory" type="add" due-to="bdavis@saintandreas.org">
+        [FTP] Configuration does not include option for setting socket timeout.
+      </action>
+      <action issue="VFS-414" dev="ggregory" type="add" due-to="ggregory">
         [FTP] Add config API to set the file type.
       </action>
-      <action issue="VFS-182" dev="ggregory" type="update" due-to="ggregory">
+      <action issue="VFS-182" dev="ggregory" type="add" due-to="ggregory">
         [FTP] Usage of FTP with heterogeneous FTP server (possibility of using Ascii file type).
       </action>
       <action issue="VFS-395" dev="ggregory" type="update" due-to="ggregory">
